app/assets/javascripts/govuk_publishing_components/analytics-ga4/ga4-form-tracker.js

```diff
@@ -0,0 +1,113 @@
+window.GOVUK = window.GOVUK || {}
+window.GOVUK.Modules = window.GOVUK.Modules || {};
+
+(function (Modules) {
+  'use strict'
+
+  function Ga4FormTracker (module) {
+    this.module = module
+    this.trackingTrigger = 'data-ga4-form' // elements with this attribute get tracked
+  }
+
+  Ga4FormTracker.prototype.init = function () {
+    var consentCookie = window.GOVUK.getConsentCookie()
+
+    if (consentCookie && consentCookie.settings) {
+      this.startModule()
+    } else {
+      this.startModule = this.startModule.bind(this)
+      window.addEventListener('cookie-consent', this.startModule)
+    }
+  }
+
+  // triggered by cookie-consent event, which happens when users consent to cookies
+  Ga4FormTracker.prototype.startModule = function () {
+    if (window.dataLayer) {
+      this.module.addEventListener('submit', this.trackFormSubmit.bind(this))
+    }
+  }
+
+  Ga4FormTracker.prototype.trackFormSubmit = function (event) {
+    var target = window.GOVUK.analyticsGa4.core.trackFunctions.findTrackingAttributes(event.target, this.trackingTrigger)
+    if (target) {
+      try {
+        var data = target.getAttribute(this.trackingTrigger)
+        data = JSON.parse(data)
+      } catch (e) {
+        // if there's a problem with the config, don't start the tracker
+        console.warn('GA4 configuration error: ' + e.message, window.location)
+        return
+      }
+
+      var formInputs = this.getFormInputs()
+      var formData = this.getInputValues(formInputs)
+      data.text = this.combineGivenAnswers(formData) || 'No answer given'
+
+      var schemas = new window.GOVUK.analyticsGa4.Schemas()
+      var schema = schemas.mergeProperties(data, 'event_data')
+      window.GOVUK.analyticsGa4.core.sendData(schema)
+    }
+  }
+
+  Ga4FormTracker.prototype.getFormInputs = function () {
+    var inputs = []
+    var labels = this.module.querySelectorAll('label')
+
+    for (var i = 0; i < labels.length; i++) {
+      var label = labels[i]
+      var labelFor = label.getAttribute('for')
+      var input = false
+      if (labelFor) {
+        input = this.module.querySelector('[id=' + labelFor + ']')
+      } else {
+        input = label.querySelector('input')
```

Comment:
If I’m understanding this correctly, this code will look for an `input` tag inside the `label` if the `for` attribute doesn’t exist. But in the smart answers HTML, it looks like the `labels` and `inputs` are siblings rather than child/parent and when I run this `else` statement locally, the `input` is getting set to `null` - is this expected behaviour?

Reply:
It should work like this - if the label has a `for` attribute, we use that to find the input. If it doesn't, either the input is inside the the label (hence the `else`) or the form is invalid, because the label isn't associated with the input. I'd assumed that our forms were valid, can you give me an example of where this isn't working?

Reply:
Oh I see. I was assuming all the labels would have a `for` attribute and, if they didn't, they would remain as sibling elements to the input (which would indeed break the association between the `label` and the `input` as you say) - I didn't come across this scenario naturally when testing, I forced it by amending the HTML (so in other words, everything is working as it should) 👍